core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -86,14 +89,69 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
+
+    // Create /controller and update /controller_epoch atomically
+    val newControllerEpoch = curEpoch + 1
+    val expectedControllerEpochZkVersion = curEpochZkVersion
+
+    debug(s"Try to create ${ControllerZNode.path} and increment controller epoch to $newControllerEpoch with expected controller epoch zkVersion $expectedControllerEpochZkVersion")
+    def tryCreateControllerZNodeAndIncrementEpoch(): (Int, Int) = {
+      try {
+        val transaction = zooKeeperClient.createTransaction()
+        transaction.create(ControllerZNode.path, ControllerZNode.encode(controllerId, timestamp),
+          acls(ControllerZNode.path).asJava, CreateMode.EPHEMERAL)
+        transaction.setData(ControllerEpochZNode.path, ControllerEpochZNode.encode(newControllerEpoch), expectedControllerEpochZkVersion)
+        val results = transaction.commit()
+        val setDataResult = results.get(1).asInstanceOf[SetDataResult]
+        (newControllerEpoch, setDataResult.getStat.getVersion)
+      } catch {
+        case _: NodeExistsException =>
+          val curControllerId = getControllerId.getOrElse(throw new ControllerMovedException(
+            s"The ephemeral node at ${ControllerZNode.path} went away while checking whether the controller election succeeds. " +
+              s"Aborting controller startup procedure"))
+          if (controllerId == curControllerId) {
+            val (epoch, stat)  = getControllerEpoch.getOrElse(throw new ControllerMovedException(
+              "Controller moved to another broker. Aborting controller startup procedure"))
+            // If the epoch is the same as newControllerEpoch, it is safe to infer that the returned epoch zkVersion
+            // is associated with the current broker during controller election because we already knew that the zk
+            // transaction succeeds based on the controller znode verification. Other rounds of controller
+            // election will result in larger epoch number written in zk.
+            if (epoch == newControllerEpoch)
+              return (newControllerEpoch, stat.getVersion)
+          }
+          throw new ControllerMovedException("Controller moved to another broker. Aborting controller startup procedure")
+        case _: BadVersionException =>
+          throw new ControllerMovedException("Controller moved to another broker. Aborting controller startup procedure")
+        case _: ConnectionLossException => tryCreateControllerZNodeAndIncrementEpoch()
+      }
+    }
+
+    tryCreateControllerZNodeAndIncrementEpoch()
+  }
+
+  private def createControllerEpochZNode(): (Int, Int) = {
+    createControllerEpochRaw(KafkaController.InitialControllerEpoch - 1).resultCode match {
+      case Code.OK =>
+        info(s"Successfully create ${ControllerEpochZNode.path} with initial epoch ${KafkaController.InitialControllerEpoch - 1}")
+        (KafkaController.InitialControllerEpoch - 1, KafkaController.InitialControllerEpochZkVersion - 1)
```

Comment:
It seems that every usage of KafkaController.InitialControllerEpoch and KafkaController.InitialControllerEpochZkVersion as 0 requires subtraction by 1. Could we just define them as 0?

Reply:
Yes. Done.